src/lib/data/tob.ts

```diff
@@ -0,0 +1,649 @@
+	for (const [key, val] of Object.entries(maxStats) as [keyof GearStats, number][]) {
+		if (val <= 0 || ignoreStats.includes(key)) continue;
+		const rawPercent = Math.min(100, calcWhatPercent(userStats[key], val));
+		totalPercent += rawPercent;
+		numKeys++;
+	}
+	// For melee compare the highest melee attack stat of max setup with the highest melee attack stat of the user
+	if (melee) {
+		let maxMeleeStat = Math.max(
+			maxStats['attack_stab'],
+			Math.max(maxStats['attack_slash'], maxStats['attack_crush'])
+		);
+		let userMeleeStat = Math.max(
+			userStats['attack_stab'],
+			Math.max(userStats['attack_slash'], userStats['attack_crush'])
+		);
+		totalPercent += Math.min(100, calcWhatPercent(userMeleeStat, maxMeleeStat));
+		numKeys++;
+	}
+	totalPercent /= numKeys;
+	// Heavy penalize for having less than 50% in the main stat of this setup.
+	if (userStats[heavyPenalizeStat] < maxStats[heavyPenalizeStat] / 2) {
+		totalPercent = Math.floor(Math.max(0, totalPercent / 2));
+	}
+	return totalPercent;
+}
+
+export const TOBMaxMageGear = constructGearSetup({
+	head: 'Ancestral hat',
+	neck: 'Occult necklace',
+	body: 'Ancestral robe top',
+	cape: 'Imbued saradomin cape',
+	hands: 'Tormented bracelet',
+	legs: 'Ancestral robe bottom',
+	feet: 'Eternal boots',
+	weapon: 'Kodai wand',
+	shield: 'Arcane spirit shield',
+	ring: 'Seers ring(i)'
+});
+const maxMage = new Gear(TOBMaxMageGear);
+
+export const TOBMaxRangeGear = constructGearSetup({
+	head: 'Void ranger helm',
+	neck: 'Necklace of anguish',
+	body: 'Elite void top',
+	cape: "Ava's assembler",
+	hands: 'Void knight gloves',
+	legs: 'Elite void robe',
+	feet: 'Pegasian boots',
+	'2h': 'Twisted bow',
+	ring: 'Archers ring(i)',
+	ammo: 'Dragon arrow'
+});
+const maxRange = new Gear(TOBMaxRangeGear);
+maxRange.ammo!.quantity = 10_000;
+
+export const TOBMaxMeleeGear = constructGearSetup({
+	head: 'Neitiznot faceguard',
+	neck: 'Amulet of torture',
+	body: 'Bandos chestplate',
+	cape: 'Infernal cape',
+	hands: 'Ferocious gloves',
+	legs: 'Bandos tassets',
+	feet: 'Primordial boots',
+	weapon: 'Abyssal tentacle',
+	ring: 'Berserker ring(i)',
+	shield: 'Avernic defender'
+});
+const maxMelee = new Gear(TOBMaxMeleeGear);
+
+export function calculateTOBUserGearPercents(user: KlasaUser) {
+	const melee = calcSetupPercent(
+		maxMelee.stats,
+		user.getGear('melee').stats,
+		'melee_strength',
+		['attack_stab', 'attack_slash', 'attack_crush', 'attack_ranged', 'attack_magic'],
+		true
+	);
+	const range = calcSetupPercent(
+		maxRange.stats,
+		user.getGear('range').stats,
+		'ranged_strength',
+		['attack_stab', 'attack_slash', 'attack_crush', 'attack_magic'],
+		false
+	);
+	const mage = calcSetupPercent(
+		maxMage.stats,
+		user.getGear('mage').stats,
+		'magic_damage',
+		['attack_stab', 'attack_slash', 'attack_crush', 'attack_ranged'],
+		false
+	);
+	return {
+		melee,
+		range,
+		mage,
+		total: (melee + range + mage) / 3
+	};
+}
+
+export const minimumCoxSuppliesNeeded = new Bank({
```

Comment:
Not necessary, but probably should rename this const to avoid confusion and potentially mis-importing